components/ai_chat/core/browser/engine/engine_consumer_oai.cc

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/browser/engine/engine_consumer_oai.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "base/containers/fixed_flat_set.h"
+#include "base/containers/flat_set.h"
+#include "base/functional/bind.h"
+#include "base/i18n/time_formatting.h"
+#include "base/memory/raw_ptr.h"
+#include "base/memory/weak_ptr.h"
+#include "base/strings/pattern.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/types/expected.h"
+#include "brave/components/ai_chat/core/browser/engine/engine_consumer.h"
+#include "brave/components/ai_chat/core/browser/engine/remote_completion_client.h"
+#include "brave/components/ai_chat/core/common/features.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom-forward.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "components/grit/brave_components_strings.h"
+#include "net/http/http_status_code.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace {
+constexpr char kQuestionPrompt[] =
+    "Propose up to 3 very short questions that a reader may ask about the "
+    "content. Wrap each in <question> tags.";
+}  // namespace
+
+namespace ai_chat {
+
+namespace {
+
+using mojom::CharacterType;
+using mojom::ConversationTurn;
+
+base::Value::List BuildMessages(
+    const std::string& human_input,
+    const std::string& page_content,
+    const std::optional<std::string>& selected_text,
+    const bool& is_video,
+    const EngineConsumer::ConversationHistory& conversation_history) {
+  const std::string prompt_segment_article =
+      page_content.empty()
+          ? ""
+          : base::StrCat(
+                {base::ReplaceStringPlaceholders(
+                     l10n_util::GetStringUTF8(
+                         is_video ? IDS_AI_CHAT_CLAUDE_VIDEO_PROMPT_SEGMENT
+                                  : IDS_AI_CHAT_CLAUDE_ARTICLE_PROMPT_SEGMENT),
+                     {page_content}, nullptr),
+                 "\n\n"});
+
+  std::string date_and_time_string =
+      base::UTF16ToUTF8(TimeFormatFriendlyDateAndTime(base::Time::Now()));
+  std::string system_message = base::StrCat(
```

Comment:
Let's not have the system message. This is not supported by all models and sometimes even if supported it causes issues for the model. Not having the system message will keep the prompting as general as possible